### ISSUE TITLE
So close, but not quite working...?

### ISSUE THREAD

**Issue by mapleroyal** [bryansteiner/gpu-passthrough-tutorial#35]
TLDR: when I run the VM, nothing seems to happen (it doesn't appear to run, and it doesn't take over my video, keyboard, or mouse), but if I log out I may not have keyboard/mouse, and if I do and I log back in then one time it showed the VM video (the very early boot phase with the tianocore logo) and I had no keyboard/mouse.

Also, THANK YOU for the guide! Every other guide on the internet wants you to lock down the dGPU so only the guest can use it which seems insane and is also 30x more complicated than the (more useful) way that you do it.

___

**SYSTEM:**
- Mobo: asus z690 extreme
- CPU: intel 13900k
- GPU: rtx 4090
- Displays:
    - one connected to the dGPU via DP
    - one connected to the iGPU via HDMI
- OS: vanilla Arch, linux-zen, x11, GNOME
    - installed via `archinstall` so everything just works™

**A couple notes:**
- VM is using windows 11 and is therefore named `win11`
- The packages I install to get virt-manager set up:
```bash
sudo pacman -S qemu virt-manager virt-viewer dnsmasq vde2 bridge-utils libguestfs dmidecode spice-vdagent
# choose qemu-desktop instead of qemu-base or qemu-full when asked which repository extras I'd like
```
- `UEFI x86_64: /usr/share/OVMF/OVMF_CODE.fd` is not an option, instead there is `UEFI x86_64: /usr/share/edk2/x64/OVMF_CODE.fd`
    - however, I've tried with that and with the regular `UEFI` and I still get the same issue
- The 4090 only has the GPU and audio to pass through
- I've tried with and without doing the CPU host-passthrough step
- In an effort to make this first setup as non-complex as possible, I'm going with standard image storage and no cpu pinning/hugepages

**This runs with no issue:**
```bash
$ sudo mkdir -p /etc/libvirt/hooks
$ sudo wget 'https://raw.githubusercontent.com/PassthroughPOST/VFIO-Tools/master/libvirt_hooks/qemu' \
     -O /etc/libvirt/hooks/qemu
$ sudo chmod +x /etc/libvirt/hooks/qemu
```

**This command**
```bash
sudo systemctl restart libvirtd
```
is this command in arch
```bash
sudo systemctl restart libvirtd
```
but it works with no issues.

**The hooks:**
```bash
$ tree /etc/libvirt/hooks/
/etc/libvirt/hooks/
├── kvm.conf
├── qemu
└── qemu.d
    └── win11
        ├── prepare
        │   └── begin
        │       └── bind_vfio.sh
        └── release
            └── end
                └── unbind_vfio.sh
```
```bash
$ cat /etc/libvirt/hooks/kvm.conf
## Virsh devices
VIRSH_GPU_VIDEO=pci_0000_01_00_0
VIRSH_GPU_AUDIO=pci_0000_01_00_1
```
```bash
$ cat /etc/libvirt/hooks/qemu.d/win11/prepare/begin/bind_vfio.sh
#!/bin/bash

## Load the config file
source "/etc/libvirt/hooks/kvm.conf"

## Load vfio
modprobe vfio
modprobe vfio_iommu_type1
modprobe vfio_pci

## Unbind gpu from nvidia and bind to vfio
virsh nodedev-detach $VIRSH_GPU_VIDEO
virsh nodedev-detach $VIRSH_GPU_AUDIO
```
```bash
$ cat /etc/libvirt/hooks/qemu.d/win11/release/end/unbind_vfio.sh
#!/bin/bash

## Load the config file
source "/etc/libvirt/hooks/kvm.conf"

## Unbind gpu from vfio and bind to nvidia
virsh nodedev-reattach $VIRSH_GPU_VIDEO
virsh nodedev-reattach $VIRSH_GPU_AUDIO

## Unload vfio
modprobe -r vfio_pci
modprobe -r vfio_iommu_type1
modprobe -r vfio
```

The rest of the steps are followed precisely as they aren't unique to my machine. I don't have issues creating VM's generally.

Any ideas what's going on here?

**Comment by bryansteiner** [comment]
Have you been able to run the `prepare/begin` hook scripts individually prior to starting your VM? If there is a hangup caused by any one of those that would explain your scenario. It seems like you're only using `bind_vfio.sh` on startup so make sure to test and confirm that you're able to run each command individually. Others have reported issues in the past with their specific setups and this script so if you need assistance, check out other issues that have been closed and report the same problem.

**Comment by bryansteiner** [comment]
Closed. No response with additional requested information.
